kuma/static/js/newsletter.js

```diff
@@ -0,0 +1,221 @@
+
+        // handle errors
+        var errorArray = [];
+        var newsletterErrors = doc.getElementById('newsletterErrors');
+        function newsletterError(event) {
+            var errorList = doc.createElement('ul');
+            errorList.className ='errorlist';
+
+            if(errorArray.length) {
+                for (var i = 0; i < errorArray.length; i++) {
+                    var item = doc.createElement('li');
+                    item.appendChild(doc.createTextNode(errorArray[i]));
+                    errorList.appendChild(item);
+                }
+            } else {
+                // no error messages, forward to server for better troubleshooting
+                newsletterForm.setAttribute('data-skip-xhr', true);
+                newsletterForm.submit();
+            }
+
+            newsletterErrors.appendChild(errorList);
+            newsletterErrors.style.display = 'block';
+
+            mdn.analytics.trackEvent({
+                'category': 'newsletter',
+                'action': 'progression',
+                'label': 'error'
+            });
+        }
+
+        // show sucess message
+        function newsletterThanks() {
+            var thanks = doc.getElementById('newsletterThanks');
+
+            // show thanks message
+            thanks.style.display = 'block';
+
+            // track success
+            mdn.analytics.trackEvent({
+                'category': 'newsletter',
+                'action': 'progression',
+                'label': 'complete'
+            });
+
+            // don't show signup form on article pages anymore
+            newsletterSaveHide();
+        }
+
+        // XHR subscribe; handle errors; display thanks message on success.
+        function newsletterSubscribe(event) {
+            var skipXHR = newsletterForm.getAttribute('data-skip-xhr');
+            if (skipXHR) {
+                // track this
+                mdn.analytics.trackEvent({
+                    'category': 'newsletter',
+                    'action': 'progression',
+                    'label': 'error-forward'
+                });
+
+                return true;
+            }
+            event.preventDefault();
+            event.stopPropagation();
+
+            // track this submission
+            mdn.analytics.trackEvent({
+                'category': 'newsletter',
+                'action': 'progression',
+                'label': 'submission'
+            });
+
+            // new submission, clear old errors
+            errorArray = [];
+            newsletterErrors.style.display = 'none';
+            while (newsletterErrors.firstChild) newsletterErrors.removeChild(newsletterErrors.firstChild);
+
+            var fmt = doc.getElementById('fmt').value;
+            var email = newsletterEmailInput.value;
+            var newsletter = doc.getElementById('newsletterNewslettersInput').value;
+            var privacy = doc.querySelector('input[name="privacy"]:checked') ? '&privacy=true' : '';
```

Comment:
When a user submits without clicking the "Privacy Policy" checkbox, the submission failed and an error event is sent to GA. Should this be handled client-side (omit the XHR), with a different error event?